arbnode/transaction_streamer.go

```diff
@@ -1167,7 +1231,123 @@ func (s *TransactionStreamer) executeMessages(ctx context.Context, ignored struc
 	return s.config().ExecuteMessageLoopDelay
 }
 
+func (s *TransactionStreamer) SubmitEspressoTransactionPos(pos arbutil.MessageIndex) {
+	s.pendingTxnsQueueMutex.Lock()
+	defer s.pendingTxnsQueueMutex.Unlock()
+	s.pendingTxnsPos = append(s.pendingTxnsPos, pos)
+}
+
+func (s *TransactionStreamer) PollSubmittedTransactionForFinality(ctx context.Context) time.Duration {
+
+	data, err := s.espressoClient.FetchTransactionByHash(ctx, s.submittedTxHash)
+	if err != nil {
+		log.Error("failed to fetch the transaction hash", "err", err, "pos", s.submittedTxnPos)
+		return s.config().EspressoTxnsPollingInterval
+	}
+	// get the message at the submitted txn position
+	msg, err := s.getMessageWithMetadataAndBlockHash(*s.submittedTxnPos)
+	if err != nil {
+		log.Error("failed to get espresso message", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+	// parse the message to get the transaction bytes and the justification
+	txns, jst, err := arbos.ParseEspressoMsg(msg.MessageWithMeta.Message)
+	if err != nil {
+		log.Error("failed to parse espresso message", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+
+	//If we reach here, the transaction has been included in a block
+	//after confirming the transaction, we need to fetch the header and fill it into the block justification.
+	espressoHeader, err := s.espressoClient.FetchHeaderByHeight(ctx, data.BlockHeight)
+	if err != nil {
+		log.Error("espresso: failed to fetch header by height ", "err", err)
+		return s.config().EspressoTxnsPollingInterval
+	}
+
+	// fetch the namespace proof and vid common. Should use a more efficient way
+	resp, err := s.espressoClient.FetchTransactionsInBlock(ctx, data.BlockHeight, s.config().EspressoNamespace)
+	if err != nil {
+		log.Info("failed to fetch the transactions in block")
```

Comment:
This sounds a bit like an error now for anyone looking at the logs. We can add something like "will retry" if the message is expected to pop up. Although I think that since we did fetch the header this should be relatively rare, so should it be WARN level instead?